ansible_collections/arista/avd/roles/eos_designs/python_modules/underlay/router_bgp.py

```diff
@@ -0,0 +1,124 @@
+            "maximum_routes": 12000,
+            "send_community": "all",
+            "struct_cfg": get(self._hostvars, "switch.bgp_peer_groups.ipv4_underlay_peers.structured_config"),
+        }
+        router_bgp["peer_groups"] = {self._peer_group_ipv4_underlay_peers_name: peer_group}
+
+        # Address Families
+        # TODO - see if it makes sense to extract logic in method
+        address_family_ipv4_peer_group = {"activate": True}
+
+        if self._underlay_rfc5549 is True:
+            address_family_ipv4_peer_group["next_hop"] = {"address_family_ipv6_originate": True}
+
+        router_bgp["address_family_ipv4"] = {
+            "peer_groups": {
+                self._peer_group_ipv4_underlay_peers_name: address_family_ipv4_peer_group,
+            }
+        }
+
+        if self._underlay_ipv6 is True:
+            router_bgp["address_family_ipv6"] = {
+                "peer_groups": {
+                    self._peer_group_ipv4_underlay_peers_name: {"activate": True},
+                }
+            }
+
+        # Redistribute routes
+        router_bgp["redistribute_routes"] = self._router_bgp_redistribute_routes
+
+        # Neighbor Interfaces
+        if self._underlay_rfc5549 is True:
+            neighbor_interfaces = {}
+            for link in self._underlay_links:
+                if link["type"] != "underlay_p2p":
+                    continue
+
+                neighbor_interface = {
+                    "peer_group": self._peer_group_ipv4_underlay_peers_name,
+                    "remote_as": link["peer_bgp_as"],
+                    "description": "_".join([link["peer"], link["peer_interface"]]),
+                }
+
+                if self._filter_peer_as is True:
+                    self._underlay_filter_peer_as_route_maps_asns.append(link["peer_bgp_as"])
+
+                neighbor_interfaces[link["interface"]] = neighbor_interface
+
+            if neighbor_interfaces:
+                router_bgp["neighbor_interfaces"] = neighbor_interfaces
+        # Neighbors
+        else:
+            neighbors = {}
+            for link in self._underlay_links:
+                if link["type"] != "underlay_p2p":
+                    continue
+
+                neighbor = {
+                    "peer_group": self._peer_group_ipv4_underlay_peers_name,
+                    "remote_as": get(link, "peer_bgp_as"),
+                    "description": "_".join([link["peer"], link["peer_interface"]]),
+                    "bfd": get(link, "bfd"),
+                }
+
+                if self._filter_peer_as is True:
+                    neighbor["route_map_out"] = f"RM-BGP-AS{link['peer_bgp_as']}-OUT"
+
+                neighbors[link["peer_ip_address"]] = neighbor
+
+            if neighbors:
+                router_bgp["neighbors"] = neighbors
+
+        # Need to keep potentially empty dict for redistribute_routes
+        return strip_empties_from_dict(router_bgp, strip_values_tuple=(None, ""))
```

Comment:
so this is actually making the empty dict inside `connected` to survive right? or are you saying it could be removed later down the line?

returning an empty dict from the method would lead to
```
redistribute_routes: {}
```

which should not survive right? 

only this:

```
redistribute_routes: {
  connected: {}
}
```

is what we want to keep - or am I missing a point?

Reply:
Basically you probably need to prune the `None` values as you build out data with optional keys, instead of waiting to the end. Also notice we have `strip_null_from_data` to do what would normally be required.

Reply:
We should not remove `""`. If someone set an input to `""` it is probably because that's what they want :)